Plugin fails to install default template when using legacy API with Elasticsearch 8.x (or higher) versions.

### Issue description

When we upgrade Elasticsearch to 8.x and use legacy template (`template_api`=>`legacy`), the plugin resolves the composable style template as a default template and send to legacy API, obviously it doesn't work (see the produced error below). 
The discussion with team brought to the idea of using legacy template api with custom template only. However, if ILM (ilm_* settings) is enabled, the plugin crashes. When plugin tries to set ILM settings to provided custom template, the plugin aligns on Elasticsearch version (pseudo: `template_settings  = version < 8 ? template['settings']: template['template']['settings']`). Obviously, with custom _flexible_ template it is hard to guess where ILM settings located.

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem,
including (e.g.) pipeline definition(s), settings, locale, etc.  The easier
you make for us to reproduce it, the more likely that somebody will take the
time to look at it.

 1. set `template_api` to `legacy`
 2. connect to any (including cloud) Elasticsearch where its version is 8.x+
 3. run Logstash and check the log

**Provide logs (if relevant)**:
```
[ERROR] 2022-12-13 23:42:54.690 [Ruby-0-Thread-12: :1] elasticsearch - Failed to install template 
{:message=>"Got response code '400' contacting Elasticsearch at URL 
'https://my-es-host.com:9200/_template/logstashb'", :exception=>LogStash::Outputs::ElasticSearch::HttpClient::Pool::BadResponseCodeError, 
:backtrace=>["/grid01/apps/PIPELINES/logstash-8.2.3/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-11.12.1-
java/lib/logstash/outputs/elasticsearch/http_client/manticore_adapter.rb:84:in `perform_request'", 
"/grid01/apps/PIPELINES/logstash-8.2.3/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-11.12.1-
java/lib/logstash/outputs/elasticsearch/http_client/pool.rb:324:in `perform_request_to_url'", "/grid01/apps/PIPELINES/logstash-8.2.3/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-11.12.1-
java/lib/logstash/outputs/elasticsearch/http_client/pool.rb:311:in `block in perform_request'", "/grid01/apps/PIPELINES/logstash-8.2.3/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-11.12.1-
java/lib/logstash/outputs/elasticsearch/http_client/pool.rb:398:in `with_connection'", "/grid01/apps/PIPELINES/logstash-8.2.3/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-11.12.1-
java/lib/logstash/outputs/elasticsearch/http_client/pool.rb:310:in `perform_request'", "/grid01/apps/PIPELINES/logstash-8.2.3/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-11.12.1-java/lib/logstash/outputs/elasticsearch/http_client/pool.rb:318:in `block in Pool'", "/grid01/apps/PIPELINES/logstash-8.2.3/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-11.12.1-java/lib/logstash/outputs/elasticsearch/http_client.rb:412:in `template_put'", "/grid01/apps/PIPELINES/logstash-8.2.3/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-11.12.1-java/lib/logstash/outputs/elasticsearch/http_client.rb:85:in `template_install'", "/grid01/apps/PIPELINES/logstash-8.2.3/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-11.12.1-java/lib/logstash/outputs/elasticsearch/template_manager.rb:41:in `install'", "/grid01/apps/PIPELINES/logstash-8.2.3/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-11.12.1-
java/lib/logstash/outputs/elasticsearch/template_manager.rb:29:in `install_template'", "/grid01/apps/PIPELINES/logstash-
8.2.3/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-11.12.1-java/lib/logstash/outputs/elasticsearch.rb:583:in 
`install_template'", "/grid01/apps/PIPELINES/logstash-8.2.3/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-11.12.1-
java/lib/logstash/outputs/elasticsearch.rb:349:in `finish_register'", "/grid01/apps/PIPELINES/logstash-
8.2.3/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-11.12.1-java/lib/logstash/outputs/elasticsearch.rb:305:in 
`block in register'", "/grid01/apps/PIPELINES/logstash-8.2.3/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-11.12.1-
java/lib/logstash/plugin_mixins/elasticsearch/common.rb:167:in `block in after_successful_connection'"]}
```

### Acceptance Criteria
- Legacy template API is used with only custom templates
- Plugin should not crash, instead provide a clear message/guide what to do next

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,6 @@
+## 11.13.1
+ - Fixes the failure when legacy template (`template_api=>'legacy'`) API is used with custom template, the plugin doesn't resolve ILM settings and crash. [#1102](https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/1102)
+
 ## 11.13.0
  - add technology preview support for allowing events to individually encode a default pipeline with `[@metadata][target_ingest_pipeline]` (as part of a technology preview, this feature may change without notice) [#1113](https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/1113)
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,5 @@
 ## 11.13.1
- - Fixes the failure when legacy template (`template_api=>'legacy'`) API is used with custom template, the plugin doesn't resolve ILM settings and crash. [#1102](https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/1102)
+ - Avoid crash by ensuring ILM settings are injected in the correct location depending on the default (or custom) template format, template_api setting and ES version [#1102](https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/1102)
 
 ## 11.13.0
  - add technology preview support for allowing events to individually encode a default pipeline with `[@metadata][target_ingest_pipeline]` (as part of a technology preview, this feature may change without notice) [#1113](https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/1113)

--- a/lib/logstash/outputs/elasticsearch/template_manager.rb
+++ b/lib/logstash/outputs/elasticsearch/template_manager.rb
@@ -54,19 +54,27 @@ module LogStash; module Outputs; class ElasticSearch
     end
 
     def self.resolve_template_settings(plugin, template)
-      return composable_index_template_settings(template) if template.key?('template')
-      return legacy_index_template_settings(template) if template.key?('settings')
-      return template_endpoint(plugin) == INDEX_TEMPLATE_ENDPOINT ?
-               composable_index_template_settings(template) :
-               legacy_index_template_settings(template)
+      if template.key?('template')
+        plugin.logger.trace("Resolving ILM template settings: under 'template' key", :template => template, :template_api => plugin.template_api, :es_version => plugin.maximum_seen_major_version)
+        composable_index_template_settings(template)
+      elsif template.key?('settings')
+        plugin.logger.trace("Resolving ILM template settings: under 'settings' key", :template => template, :template_api => plugin.template_api, :es_version => plugin.maximum_seen_major_version)
+        legacy_index_template_settings(template)
+      else
+        template_endpoint = template_endpoint(plugin)
+        plugin.logger.trace("Resolving ILM template settings: template doesn't have 'settings' or 'template' fields, falling back to auto detection", :template => template, :template_api => plugin.template_api, :es_version => plugin.maximum_seen_major_version, :template_endpoint => template_endpoint)
+        template_endpoint == INDEX_TEMPLATE_ENDPOINT ?
+          composable_index_template_settings(template) :
+          legacy_index_template_settings(template)
+      end
     end
 
-    # Returns (if exists) or creates _template API compatible template settings
+    # Sets ['settings'] field to be compatible with _template API structure
     def self.legacy_index_template_settings(template)
       template['settings'] ||= {}
     end
 
-    # Returns (if exists) or creates _index_template API compatible template settings
+    # Sets the ['template']['settings'] fields if not exist to be compatible with _index_template API structure
     def self.composable_index_template_settings(template)
       template['template'] ||= {}
       template['template']['settings'] ||= {}

--- a/lib/logstash/outputs/elasticsearch/template_manager.rb
+++ b/lib/logstash/outputs/elasticsearch/template_manager.rb
@@ -46,15 +46,30 @@ module LogStash; module Outputs; class ElasticSearch
       # definition - remove any existing definition of 'template'
       template.delete('template') if template.include?('template') if plugin.maximum_seen_major_version < 8
       template['index_patterns'] = "#{plugin.ilm_rollover_alias}-*"
-      settings = template_settings(plugin, template)
+      settings = resolve_template_settings(plugin, template)
       if settings && (settings['index.lifecycle.name'] || settings['index.lifecycle.rollover_alias'])
         plugin.logger.info("Overwriting index lifecycle name and rollover alias as ILM is enabled")
       end
       settings.update({ 'index.lifecycle.name' => plugin.ilm_policy, 'index.lifecycle.rollover_alias' => plugin.ilm_rollover_alias})
     end
 
-    def self.template_settings(plugin, template)
-      plugin.maximum_seen_major_version < 8 ? template['settings']: template['template']['settings']
+    def self.resolve_template_settings(plugin, template)
+      return composable_index_template_settings(template) if template.key?('template')
+      return legacy_index_template_settings(template) if template.key?('settings')
+      return template_endpoint(plugin) == INDEX_TEMPLATE_ENDPOINT ?
+               composable_index_template_settings(template) :
+               legacy_index_template_settings(template)
+    end
+
+    # Returns (if exists) or creates _template API compatible template settings
+    def self.legacy_index_template_settings(template)
+      template['settings'] ||= {}
+    end
+
+    # Returns (if exists) or creates _index_template API compatible template settings
+    def self.composable_index_template_settings(template)
+      template['template'] ||= {}
+      template['template']['settings'] ||= {}
     end
 
     # Template name - if template_name set, use it

--- a/logstash-output-elasticsearch.gemspec
+++ b/logstash-output-elasticsearch.gemspec
@@ -1,6 +1,6 @@
 Gem::Specification.new do |s|
   s.name            = 'logstash-output-elasticsearch'
-  s.version         = '11.13.0'
+  s.version         = '11.13.1'
   s.licenses        = ['apache-2.0']
   s.summary         = "Stores logs in Elasticsearch"
   s.description     = "This gem is a Logstash plugin required to be installed on top of the Logstash core pipeline using $LS_HOME/bin/logstash-plugin install gemname. This gem is not a stand-alone program"

--- a/spec/unit/outputs/elasticsearch/template_manager_spec.rb
+++ b/spec/unit/outputs/elasticsearch/template_manager_spec.rb
@@ -73,6 +73,7 @@ describe LogStash::Outputs::ElasticSearch::TemplateManager do
         let(:template_api) { "composable" }
 
         it 'resolves composable index template API compatible setting' do
+          expect(plugin).to receive(:maximum_seen_major_version).at_least(:once).and_return(8) # required to log
           template = {}
           described_class.resolve_template_settings(plugin, template)
           expect(template["template"]["settings"]).not_to eq(nil)
@@ -83,6 +84,7 @@ describe LogStash::Outputs::ElasticSearch::TemplateManager do
         let(:template_api) { "legacy" }
 
         it 'resolves legacy index template API compatible setting' do
+          expect(plugin).to receive(:maximum_seen_major_version).at_least(:once).and_return(7) # required to log
           template = {}
           described_class.resolve_template_settings(plugin, template)
           expect(template["settings"]).not_to eq(nil)

--- a/spec/unit/outputs/elasticsearch/template_manager_spec.rb
+++ b/spec/unit/outputs/elasticsearch/template_manager_spec.rb
@@ -1,4 +1,4 @@
-require "logstash/devutils/rspec/spec_helper"
+require_relative "../../../../spec/spec_helper"
 require "logstash/outputs/elasticsearch/template_manager"
 
 describe LogStash::Outputs::ElasticSearch::TemplateManager do
@@ -33,33 +33,83 @@ describe LogStash::Outputs::ElasticSearch::TemplateManager do
     end
   end
 
-  describe "index template with ilm settings" do
+  context "index template with ilm settings" do
     let(:plugin_settings) { {"manage_template" => true, "template_overwrite" => true} }
     let(:plugin) { LogStash::Outputs::ElasticSearch.new(plugin_settings) }
 
-    describe "in version 8+" do
-      let(:file_path) { described_class.default_template_path(8) }
-      let(:template) { described_class.read_template_file(file_path)}
+    describe "with custom template" do
 
-      it "should update settings" do
-        expect(plugin).to receive(:maximum_seen_major_version).at_least(:once).and_return(8)
-        described_class.add_ilm_settings_to_template(plugin, template)
-        expect(template['template']['settings']['index.lifecycle.name']).not_to eq(nil)
-        expect(template['template']['settings']['index.lifecycle.rollover_alias']).not_to eq(nil)
-        expect(template.include?('settings')).to be_falsey
+      describe "in version 8+" do
+        let(:file_path) { described_class.default_template_path(8) }
+        let(:template) { described_class.read_template_file(file_path)}
+
+        it "should update settings" do
+          expect(plugin).to receive(:maximum_seen_major_version).at_least(:once).and_return(8)
+          described_class.add_ilm_settings_to_template(plugin, template)
+          expect(template['template']['settings']['index.lifecycle.name']).not_to eq(nil)
+          expect(template['template']['settings']['index.lifecycle.rollover_alias']).not_to eq(nil)
+          expect(template.include?('settings')).to be_falsey
+        end
+      end
+
+      describe "in version < 8" do
+        let(:file_path) { described_class.default_template_path(7) }
+        let(:template) { described_class.read_template_file(file_path)}
+
+        it "should update settings" do
+          expect(plugin).to receive(:maximum_seen_major_version).at_least(:once).and_return(7)
+          described_class.add_ilm_settings_to_template(plugin, template)
+          expect(template['settings']['index.lifecycle.name']).not_to eq(nil)
+          expect(template['settings']['index.lifecycle.rollover_alias']).not_to eq(nil)
+          expect(template.include?('template')).to be_falsey
+        end
       end
     end
 
-    describe "in version < 8" do
-      let(:file_path) { described_class.default_template_path(7) }
-      let(:template) { described_class.read_template_file(file_path)}
+    context "resolve template setting" do
+      let(:plugin_settings) { super().merge({"template_api" => template_api}) }
 
-      it "should update settings" do
-        expect(plugin).to receive(:maximum_seen_major_version).at_least(:once).and_return(7)
-        described_class.add_ilm_settings_to_template(plugin, template)
-        expect(template['settings']['index.lifecycle.name']).not_to eq(nil)
-        expect(template['settings']['index.lifecycle.rollover_alias']).not_to eq(nil)
-        expect(template.include?('template')).to be_falsey
+      describe "with composable template API" do
+        let(:template_api) { "composable" }
+
+        it 'resolves composable index template API compatible setting' do
+          template = {}
+          described_class.resolve_template_settings(plugin, template)
+          expect(template["template"]["settings"]).not_to eq(nil)
+        end
+      end
+
+      describe "with legacy template API" do
+        let(:template_api) { "legacy" }
+
+        it 'resolves legacy index template API compatible setting' do
+          template = {}
+          described_class.resolve_template_settings(plugin, template)
+          expect(template["settings"]).not_to eq(nil)
+        end
+      end
+
+      describe "with `template_api => 'auto'`" do
+        let(:template_api) { "auto" }
+
+        describe "with ES < 8 versions" do
+
+          it 'resolves legacy index template API compatible setting' do
+            expect(plugin).to receive(:maximum_seen_major_version).at_least(:once).and_return(7)
+            template = {}
+            described_class.resolve_template_settings(plugin, template)
+            expect(template["settings"]).not_to eq(nil)
+          end
+        end
+
+        describe "with ES >= 8 versions" do
+          it 'resolves composable index template API compatible setting' do
+            expect(plugin).to receive(:maximum_seen_major_version).at_least(:once).and_return(8)
+            template = {}
+            described_class.resolve_template_settings(plugin, template)
+            expect(template["template"]["settings"]).not_to eq(nil)
+          end
+        end
       end
     end
   end

